conjure-runtime/src/service/raw/js.rs

```diff
@@ -0,0 +1,226 @@
+        Ok(RawClient(()))
+    }
+}
+
+impl Service<Request<RawRequestBody>> for RawClient {
+    type Response = Response<RawResponseBody>;
+    type Error = Box<dyn error::Error + Sync + Send>;
+
+    // The fetch API promises to call these futures sequentially
+    #[allow(clippy::await_holding_refcell_ref)]
+    async fn call(&self, req: Request<RawRequestBody>) -> Result<Self::Response, Self::Error> {
+        let (parts, body) = req.into_parts();
+
+        let init = RequestInit::new();
+        init.set_method(parts.method.as_str());
+
+        let headers = Headers::new().map_err(JsError::new)?;
+        for (mut name, value) in &parts.headers {
+            if name == header::USER_AGENT {
+                name = &FETCH_USER_AGENT;
+            }
+
+            headers
+                .append(name.as_str(), value.to_str().map_err(|e| e.to_string())?)
+                .map_err(JsError::new)?;
+        }
+
+        init.set_headers(headers.as_ref());
+
+        // We need to keep the closure alive until we finish processing the request
+        let mut pull: Option<Closure<dyn FnMut(ReadableStreamDefaultController) -> Promise>> = None;
+
+        if !body.is_end_stream() {
+            let underlying_source = UnderlyingSource::new();
+
+            let body = Rc::new(RefCell::new(body));
+            let pull = pull.insert(Closure::new(
+                move |controller: ReadableStreamDefaultController| {
+                    wasm_bindgen_futures::future_to_promise({
+                        let body = body.clone();
+                        async move {
+                            match body.borrow_mut().frame().await {
+                                Some(Ok(frame)) => match frame.data_ref() {
+                                    Some(data) => {
+                                        let chunk = Uint8Array::new_with_length(data.len() as u32);
+                                        chunk.copy_from(data);
+                                        controller.enqueue_with_chunk(&chunk.into())?;
+                                        Ok(JsValue::UNDEFINED)
+                                    }
+                                    None => {
+                                        Err(js_sys::Error::new("unsupported trailers frame").into())
+                                    }
+                                },
+                                None => {
+                                    controller.close()?;
+                                    Ok(JsValue::UNDEFINED)
+                                }
+                                Some(Err(e)) => Err(js_sys::Error::new(&e.to_string()).into()),
+                            }
+                        }
+                    })
+                },
+            ));
+            underlying_source.set_pull(pull.as_ref().unchecked_ref());
+
+            let stream = ReadableStream::new_with_underlying_source(underlying_source.as_ref())
```

Comment:
Theoretically the request and response stream handling could be made more efficient by using the BYOB APIs but I don't know if it's worth the extra complexity.